src/main/java/picard/sam/SAMSplitter.java

```diff
@@ -0,0 +1,183 @@
+    static final String USAGE_SUMMARY = "Splits a SAM or BAM file to multiple BAMs.";
+    static final String USAGE_DETAILS = "This tool splits the input SAM/BAM file into multiple BAM files " +
+            "This can be used to split a large unmapped BAM in order to parallelize alignment."+
+            "<br />" +
+            "<h4>Usage example:</h4>" +
+            "<pre>" +
+            "java -jar picard.jar SAMSplitter \\<br />" +
+            "     I=paired_unmapped_input.bam \\<br />" +
+            "     OUTPUT_DIR=out_dir \\ <br />" +
+            "     TOTAL_READS_IN_INPUT=800000000" +
+            "     SPLIT_TO_N_TEMPLATES=24000000" +
+            "</pre>" +
+            "<hr />";
+    @Argument(doc = "Input SAM/BAM file to split", shortName = StandardOptionDefinitions.INPUT_SHORT_NAME)
+    public File INPUT;
+
+    @Argument(shortName = "N", doc = "Split to have approximately N templates per output file.")
+    public int SPLIT_TO_N_TEMPLATES;
+
+    @Argument(shortName = "TOTAL_READS", doc = "Total number of reads in the input file.")
+    public int TOTAL_READS_IN_INPUT;
+
+    @Argument(shortName = "ODIR", doc = "Directory in which to output the split BAM files.")
+    public File OUTPUT_DIR;
+
+    @Argument(shortName = "PAIR", doc = "If true the input bam contains paired reads and " +
+            "each template is assumed to contain two reads (rather than one for unpaired inputs).")
+    public boolean PAIRED_INPUT = true;
+
+    private final Log log = Log.getInstance(SAMSplitter.class);
+
+    public static void main(final String[] argv) {
+        System.exit(new SAMSplitter().instanceMain(argv));
+    }
+
+    protected int doWork() {
+        IOUtil.assertFileIsReadable(INPUT);
+        final SamReader reader = SamReaderFactory.makeDefault().referenceSequence(REFERENCE_SEQUENCE).open(INPUT);
+        final SAMFileHeader header = reader.getFileHeader();
+        if (header.getSortOrder() == SAMFileHeader.SortOrder.coordinate) {
+            throw new PicardException("Cannot split coordinate sorted bam. Only queryname sorted (or querygrouped) " +
+                    "input will ensure that templates are not split into multiple files.");
+        }
+
+        final Map<String, SAMRecord> firstSeenMates = new HashMap<>();
+        final SAMFileWriterFactory factory = new SAMFileWriterFactory();
+        double unitConverter = PAIRED_INPUT ? 2.0 : 1.0;
+
+        int splitToNFiles = (int) Math.round((TOTAL_READS_IN_INPUT / unitConverter) / SPLIT_TO_N_TEMPLATES);
+        if (splitToNFiles < 1) {
+            throw new PicardException("Splitting " + TOTAL_READS_IN_INPUT + " reads into " + SPLIT_TO_N_TEMPLATES +
+                    " templates per file would result in fewer than 1 output file.");
+        }
+
+        int templatesPerFile = (int) Math.ceil((TOTAL_READS_IN_INPUT / (double) splitToNFiles) / unitConverter);
+        SAMFileWriter[] writers = generateWriters(factory, header, splitToNFiles);
```

Comment:
I stopped looking at this further as I think you need another pass based on @yfarjoun and my suggestions.